Max Health and Mana Manipulation API

absoluteAquarian opened the pull request:
### What is the new feature?
This PR aims to make modifying the player's maximum health and mana, as well as their displays, more friendly to mod developers.

Features include:
- a proper API for modifying `Player.statLifeMax` and `Player.statManaMax`
- being able to hook into the vanilla drawing for the hearts, stars and bars in the vanilla resource display sets via `ModResourceOverlay`
- allowing mod developers to define their own display sets, which can be chosen from the Settings menu, via `ModResourceDisplaySet`
  - choosing `ModResourceDisplaySet` displays is handled via the pre-existing option in the ingame Settings>Video menu
  - consequently, modded display sets are also saved to `config.json`
- removal of any code which limited the values of `Player.statLifeMax` and `Player.statManaMax`
- several new properties in `Player` for tracking how many Life Crystals, Life Fruit and Mana Crystals have been consumed
  - `Player.ConsumedLifeCrystals`
  - `Player.ConsumedLifeFruit`
  - `Player.ConsumedManaCrystals`
- a new helper method for increasing the player's health/mana and displaying the combat text via `Player.UseHealthMaxIncreasingItem()` and `Player.UseManaMaxIncreasingItem()`

### Why should this be part of tModLoader?
Modifying the player's maximum health and mana currently requires using `Player.statLifeMax2` and `Player.statManaMax2`, which -- even though it works -- is clunky and goes against the purpose of those fields: "temporary" increases/decreases.

Furthermore, modifying how vanilla's hearts/stars/bars for health and mana are drawn is cumbersome, requiring a method detour or IL edit to be used, which can cause conflicts when multiple mods try to do so at once.

As a side note, any legacy usage of modifying `Player.statLifeMax2` and `Player.statManaMax2` will still be supported, but mod developers should be encouraged to move to the new API.

### Are there alternative designs?
n/a

### Sample usage for the new feature

Modifying the player's maximum health:
```cs
// source:  Common/Players/ExampleStatIncreasePlayer.cs
// NOTE: The full source contains code for handling increasing the player's max mana as well.
//       It has been reduced for brevity's sake.

public int exampleLifeFruits;

public override void ModifyMaxStats(out StatModifier health, out StatModifier mana) {
	health = StatModifier.Default;
	health.Base = exampleLifeFruits * ExampleLifeFruit.LifePerFruit;
	// Alternatively:  health = StatModifier.Default with { Base = exampleLifeFruits * ExampleLifeFruit.LifePerFruit };
	mana = StatModifier.Default;
}
```

Drawing custom hearts over the hearts in the Classic resource display set:
```cs
// source:  Common/UI/ResourceOverlay/VanillaLifeOverlay.cs
// NOTE: The full source contains code for handling the Fancy and Bars display sets.
//       It has been reduced for brevity's sake.

public override void PostDrawResource(ResourceOverlayDrawContext context) {
	// Hearts are replaced in groups of two
	if (context.resourceNumber > 2 * Main.LocalPlayer.GetModPlayer<ExampleLifeFruitPlayer>().exampleLifeFruits)
		return;

	Asset<Texture2D> asset = context.texture;

	// "context" contains information used to draw the resource
	// If you want to draw directly on top of the vanilla hearts, just replace the texture and have the context draw the new texture
	if (asset == TextureAssets.Heart || asset == TextureAssets.Heart2) {
		context.texture = ModContent.Request<Texture2D>("ExampleMod/Common/UI/ResourceOverlay/ClassicHeartOverlay", AssetRequestMode.ImmediateLoad);
		context.Draw();
	}
}
```

### ExampleMod updates
- `Content/Items/Consumables/ExampleLifeFruit.cs` was updated to use the new API
- added `Content/Items/Consumables/ExampleManaCrystal.cs`, which serves as an example mana increasing item
- added the `Common/UI/ResourceOverlay` folder, which contains examples of modifying vanilla's resource display sets
- added the `Common/UI/ExampleDisplaySets` folder, which contains an example of using `ModResourceDisplaySet`
- `ExampleLifeFruitPlayer` was renamed to `ExampleStatIncreasePlayer` and was moved to `Common/Players`
- updated the netcode handling in `ExampleMod.Networking.cs` to account for the new `ExampleManaCrystal` stats
- updated the `en-US.hjson` localization file to add an entry for `ExampleReversedBarsDisplay` (the `ModResourceDisplaySet` example)

### Additional Information
- the netcode handling in `ModLoader/Default/ModLoaderMod.cs` was modified to be more friendly to future additions
  - consequently, the netcode for `ModAccessorySlotPlayer` was also modified
- a lot of vanilla code that was checking `Player.statLifeMax` and/or `Player.statManaMax` were changed to check the `Player.ConsumedLifeCrystals`, `Player.ConsumedLifeFruit` and `Player.ConsumedManaCrystals` properties instead, since that's what was really happening in the first place
- added two new `AmountOfLifeHearts` and `AmountOfManaStars` properties to `PlayerStatsSnapshot` and changed the `LifePerSegment` and `ManaPerSegment` fields to be getter properties dependent on these new properties
- `PlayerResourceSetsManager` (aka `Main.ResourceSetsManager`) was expanded to include `ModResourceDisplaySet` instances
  - assuming that the `ModResourceDisplaySet` was loaded, you can get the instance via `Main.ResourceSetsManager.GetDisplaySet<T>()`
  - you can also get the instance via `Main.ResourceSetsManager.GetDisplaySet("ModName/Name")`
- usage of `IPlayerResourceDisplaySet.NameKey` was removed in favor of a `DisplayedName` getter property, though the property still exists for whoever might want to use it

absoluteAquarian committed:
redid the branch due to rebase clobbering

- initial changes to the health/mana API
* added a helper function in Player for using maximum health- and mana-increasing items
* modified the AchivementHelper code in the event that achievements are enabled in the future
* added fields in Player indicating how many Life Crystals, Life Fruit and Mana Crystals they have consumed
* updated ExampleMod to account for the new API
! This PR is not complete.  UI still needs to be addressed as well as some other things.

- modified ExampleLifeFruit to better mimic how Life Fruits work

- better comment in ExampleLifeFruit

- better comment 2

absoluteAquarian committed:
added API for modifying the player's max health/mana stats

- also modified all vanilla uses of checking if statLifeMax or statManaMax were above/below a certain number with checks for consumed item counts in order to properly account for the new API
- updated ExampleLifeFruitPlayer to account for the new API

absoluteAquarian committed:
removed the added hooks to GlobalItem due to a better alternative being used

absoluteAquarian committed:
removed unnecessary parameters

absoluteAquarian committed:
initial work toward adding the API for resource UI customization

absoluteAquarian committed:
divide by 0 safeguards

absoluteAquarian committed:
added requested changes + other changes

- removed ModifyBaseMaxStats, PreModifyMaxStats and PostModifyMaxStats in favor of a StatModifier hook
- made the setters for ConsumedLifeCrystals, ConsumedLifeFruit and ConsumedManaCrystals public for better API accessibility (a mod may want to just set these directly)
- added code to save/load ConsumedLifeCrystals, ConsumedLifeFruit and ConsumedManaCrystals

absoluteAquarian committed:
further improvements

- changed ExampleLifeFruit's usage of ModifyMaxStats to be less confusing
- moved duplicated code to a PlayerLoader helper method, ResetMaxStatsToVanilla
- added Utils.Clamp usage
- clamped the values for the ConsumedLifeCrystals, ConsumeLifeFruit and ConsumedManaCrystals properties
- added/updated XML summaries

absoluteAquarian committed:
commits for hooking into Classic resource drawing (Fancy and Bars are still TODO)

- also removed the sets added to ItemID due to them not having a use
- updated ExampleMod with examples of drawing with an overlay for the Classic display set
- fixed a copy/paste typo

absoluteAquarian committed:
removed legacy code from earlier in the PR; tweaked how the Amount properties in PlayerStatsSnapshot work

absoluteAquarian committed:
XML summary fix

absoluteAquarian committed:
added abstraction and contexts

- all existing `ModResourceOverlay` hooks were converted into more "general" Resource hooks
- added `IResourceDrawSource` to `ResourceOverlayDrawContext`
- updated ExampleMod
- updated XML summaries

absoluteAquarian committed:
added Fancy and Bars displays sets support

- added missing `IResourceDrawSource` classes for the panel drawing
- fixed a bug in `PlayerIO` which prevented the ConsumedX properties in Player to save properly
- deserializing `null` as a `TagCompound` actually results in an empty tag, so i added a proper check in `PlayerIO` for when the fallback system to load the ConsumedX properties should be used
- `ModResourceOverlay` and `ResourceOverlayLoader` API changes

absoluteAquarian committed:
added ConsumeItem usage to ExampleLifeFruit to prevent consumption after the max amount of fruits are used

absoluteAquarian committed:
updated now-outdated comment in ExampleLifeFruit

absoluteAquarian committed:
API changes

- the `AmountOfHearts` property in `PlayerStatsSnapshot` was renamed to `AmountOfLifeHearts`
- added helper methods to `ResourceOverlayLoader` for drawing resources for modder convenience

absoluteAquarian committed:
removed some serious overengineering

absoluteAquarian committed:
purple tinting made the details disappear on the inventory sprite for ExampleLifeFruit + added the sprite for ExampleManaCrystal

absoluteAquarian committed:
proper application of "using more ExampleLifeFruit only plays the animation and sound"

absoluteAquarian committed:
added proper syncing of the ConsumedX properties; changed ModLoaderMod.HandlePacket to be more friendly to additional cases in the future

absoluteAquarian committed:
ExampleMod changes

- made ExampleLifeFruit actually use its texture
- renamed VanillaHealthOverlay to VanillaLifeOverlay
- renamed the "ClassicHealthOverlay" image to "ClassicLifeOverlay"
- renamed the "FancyHealthOverlay_X" images to "FancyLifeOverlay_X"
- added ExampleManaCrystal
- added VanillaManaOverlay
- renamed ExampleLifeFruitPlayer to ExampleStatIncreasePlayer and moved it to `Common/Players`
- renamed ExampleLifeFruitSystem to ExampleStatIncreaseSystem and moved it to `Common/Systems`
- made VanillaLifeOverlay "replace" the bars from right to left
- fixed some bugs caused by `float` precision limits

absoluteAquarian committed:
added ModResourceDisplaySet

- properly implemented PreDrawResourceDisplay and PostDrawResourceDisplay in the Fancy and Bars display sets
- expanded PlayerResourceSetsManager to include modded display sets
- updated XML summaries

absoluteAquarian committed:
added proper support for modded display sets in IngameOptions + example for ModResourceDisplaySet

absoluteAquarian committed:
fixed XML summaries in ModResourceDisplaySet

absoluteAquarian committed:
wrong property usage

absoluteAquarian committed:
fixed the resource set config key being improperly tracked

absoluteAquarian committed:
why was this struct `partial`?  i forgot

absoluteAquarian committed:
cleanup

absoluteAquarian committed:
code reorganization

- moved the code reponsible for adding the modded display sets to `Main.ResourceSetsManager` to before `Mod.SetupContent()` is called so that they're available for use in `Mod.PostSetupContent()` and onwards
- made the Classic display set be "registered" after the Bars display set
- changed the logic of `PlayerResourceSetsManager.CycleResourceSet()` to go from first to last instead of last to first

absoluteAquarian committed:
added more methods to `ResourceOverlayLoader` and `ModResourceDisplaySet` for "direct" drawing purposes

- sometimes a mod dev just wants to draw the vanilla thing without letting other mods affect the draw

absoluteAquarian committed:
reverted the "direct" drawing helper methods since that was a bad idea

absoluteAquarian committed:
code cleanup

absoluteAquarian committed:
XML summary cleanup

absoluteAquarian committed:
method cleanup + added helper methods for drawing the life/mana panels for the Bars display set

Chicken-Bones commented on ExampleMod/Common/Players/ExampleStatIncreasePlayer.cs:
```diff
@@ -0,0 +1,42 @@
+
+		public override void SyncPlayer(int toWho, int fromWho, bool newPlayer) {
+			ModPacket packet = Mod.GetPacket();
+			packet.Write((byte)ExampleMod.MessageType.ExamplePlayerSyncPlayer);
```
rename to `ExampleStatIncreasePlayerSync`

Chicken-Bones commented on ExampleMod/Common/Players/ExampleStatIncreasePlayer.cs:
```diff
@@ -0,0 +1,42 @@
+			// Alternatively:  mana = StatModifier.Default with { Base = exampleManaCrystals * ExampleManaCrystal.ManaPerCrystal };
+		}
+
+		public override void SyncPlayer(int toWho, int fromWho, bool newPlayer) {
```
Also need to implement `clientClone` and `SendClientChanges` to avoid desync when a player eats some mid-game

Chicken-Bones commented on ExampleMod/Common/Systems/ExampleStatIncreaseSystem.cs:
```diff
@@ -0,0 +1,29 @@
+			}
+
+			// Every odd number should display an extra star/bar compared to the number before it
+			int extraStarsToDraw = manaCrystals % 2 == 1 ? 1 : 0;
```
I don't think this is a particularly good example, it seems confusing

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/ResourceSets/ResourceDrawSettings.cs.patch:
```diff
@@ -0,0 +1,66 @@
+ 		public int ElementCount;
+ 		public int ElementIndexOffset;
+ 		public TextureGetter GetTextureMethod;
++		public ResourceOverlayDelegate GetResourceDrawContext;
```
Why does this need to be a delegate?

absoluteAquarian replied:
it doesn't have to be, but it was the simplest way to implement it if a modder wanted to use `ResourceDrawSettings` in, say, their modded display sets

absoluteAquarian replied:
edit: decided to add fields to the type instead

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ResourceDrawSources.cs:
```diff
@@ -0,0 +1,97 @@
+{
+	public abstract class ResourceDrawSource : IResourceDrawSource
+	{
+		public string Context { get; protected set; }
```
no point of a context param which is always null.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ResourceDrawSources.cs:
```diff
@@ -0,0 +1,97 @@
+
+namespace Terraria.ModLoader
+{
+	public abstract class ResourceDrawSource : IResourceDrawSource
```
Do we need this class? What's wrong with just the `DisplaySet`?

Do we need all the sub-classes?

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/IResourceDrawSource.cs:
```diff
@@ -0,0 +1,13 @@
+
+		IPlayerResourcesDisplaySet DisplaySet { get; }
+	}
+}
```
don't make an interface when only one class implements it

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModResourceDisplaySet.cs:
```diff
@@ -0,0 +1,573 @@
+		/// <param name="scale">The scale to draw the life at.  Defaults to <see cref="Vector2.One"/></param>
+		/// <param name="pulse">Whether the life's scale should be modified by the "pulse" effect.  Defaults to <see langword="false"/></param>
+		/// <returns><see langword="null"/> if the parameters are invalid, otherwise the created context</returns>
+		public static ResourceOverlayDrawContext? GetClassicHeart(SpriteBatch spriteBatch, int heartNumber, Vector2 position, Rectangle? sourceFrame = null, Color? color = null, float rotation = 0, Vector2? origin = null, Vector2? scale = null, bool pulse = false) {
```
These static helpers look really hard to maintain, and aren't used by `ExampleReversedBarsDisplay`

Chicken-Bones commented on ExampleMod/Common/UI/ExampleDisplaySets/ExampleReversedBarsDisplay.cs:
```diff
@@ -0,0 +1,222 @@
+		}
+
+		private void LifePanelContext(int elementIndex, int firstElementIndex, int lastElementIndex, out PlayerStatsSnapshot snapshot, out IResourceDrawSource drawSource) {
+			snapshot = preparedSnapshot;
```
Not a fan of this boilerplate. tML should handle this since it's exactly the same in every implementation

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/ResourceSets/PlayerStatsSnapshot.cs.patch:
```diff
@@ -0,0 +1,75 @@
++
+ namespace Terraria.GameContent.UI.ResourceSets
+ {
+ 	public struct PlayerStatsSnapshot
```
I'm not convinced of the value of this feature

absoluteAquarian replied:
`PlayerStatsSnapshot` is a vanilla type that i modified to make setting how many heart/stars/bars draw easier  
otherwise, modders would have to manually calculate the stars to draw instead of just doing something like `snapshot.AmountOfManaStars++;`

absoluteAquarian committed:
implemented requested changes

- texture swapping with `ResourceOverlayDrawContext` is no longer possible
- any methods that had a `ref ResourceOverlayDrawContext` parameter now have a parameter without the `ref`
- removed `IResourceDrawSource` in favor of `ResourceSetSlotId<T>`
- removed `ExampleStatIncreaseSystem.cs` in ExampleMod
- added more netcode to `ExampleStatIncreasePlayer.cs` in ExampleMod
- modified ExampleMod's example resource overlays and example resource display set in accordance with the new changes
- removed the helper methods in `ModResourceDisplaySet`
- moved a lot of added code in vanilla types to a corresponding `X.TML.cs` file to reduce patch sizes
- modified the changes in `ResourceDrawSettings` to be better

absoluteAquarian committed:
updated the ExampleMod resource displays with checks for vanilla textures

absoluteAquarian committed:
forgot to use SyncPlayer in `ConsumedStatIncreasesPlayer`

Chicken-Bones committed:
Merge branch '1.4' into 1.4-health-mana-api

Chicken-Bones commented on ExampleMod/Common/Players/ExampleStatIncreasePlayer.cs:
```diff
@@ -0,0 +1,61 @@
+				packet.Write((byte)Player.whoAmI);
+				packet.Write((byte)exampleLifeFruits);
+				packet.Write((byte)exampleManaCrystals);
+				packet.Send(ignoreClient: Player.whoAmI);
```
duplicate code with `SyncPlayer` above, consider common function, or just call `SyncPlayer` with `toWho: -1`

Chicken-Bones commented on ExampleMod/Common/UI/ExampleDisplaySets/ExampleReversedBarsDisplay.cs:
```diff
@@ -0,0 +1,203 @@
+		}
+
+		// DrawLife runs before DrawMana, so keep that in mind
+		// In this case, the bars are swapped, so the order of drawing (top bar, then bottom bar) should be preserved
+		// This is why Mana resources are being drawn in DrawLife and Life resources are being drawn in DrawMana
```
why must the order be preserved (top bar then bottom bar)?

absoluteAquarian replied:
otherwise, the bottom bar would clip into the top bar, which wouldn't be consistent with how the original display that this was "cloned" from works

absoluteAquarian replied:
if it would make sense for the life bar to still draw below the mana bar, then i'll change this to reflect that

Chicken-Bones replied:
I just.. can't conscience an example where `DrawLife` doesn't draw life :|

absoluteAquarian replied:
fair enough.  i'll swap the two method bodies

Chicken-Bones commented on ExampleMod/Common/UI/ResourceDisplay/VanillaLifeOverlay.cs:
```diff
@@ -0,0 +1,124 @@
+				string barsFolder = "Images/UI/PlayerResourceSets/HorizontalBars/";
+
+				// NOTE: CompareAssetNames is defined below this method's body
+				if (name == TextureAssets.Heart.Name || name == TextureAssets.Heart2.Name) {
```
compare assets directly using `==`, don't use `.Name`

absoluteAquarian replied:
fair enough.  you had wanted me to use `.Name` beforehand, but this is likely to be more reliable

Chicken-Bones commented on ExampleMod/Common/UI/ResourceDisplay/VanillaLifeOverlay.cs:
```diff
@@ -0,0 +1,124 @@
+			if (!canReplaceResource)
+				return;
+
+			// "context" contains information used to draw the resource
+			// If you want to draw directly on top of the vanilla hearts, just replace the texture and have the context draw the new texture
+			if (context.IsSlot(ClassicPlayerResourcesDisplaySet.Hearts) || context.IsSlot(FancyClassicPlayerResourcesDisplaySet.HeartsFirstRow) || context.IsSlot(FancyClassicPlayerResourcesDisplaySet.HeartsSecondRow)) {
+				// Draw over the Classic / Fancy hearts
+				DrawClassicFancyOverlay(context);
+			} else if (context.IsSlot(FancyClassicPlayerResourcesDisplaySet.HeartPanelsFirstRow) || context.IsSlot(FancyClassicPlayerResourcesDisplaySet.HeartPanelsSecondRow)) {
+				// Draw over the Fancy heart panels
+				DrawFancyPanelOverlay(context);
+			} else if (context.IsSlot(HorizontalBarsPlayerReosurcesDisplaySet.LifePanels) && fillResourceNumber >= 1 && fillResourceNumber <= context.snapshot.AmountOfLifeHearts) {
+				// Draw over the Bars middle life panels
+				DrawBarsPanelOverlay(context);
+			} else if (context.IsSlot(HorizontalBarsPlayerReosurcesDisplaySet.LifeBars)) {
+				// Draw over the Bars life bars
+				DrawBarsOverlay(context);
```
As nice as it is to showcase `IsSlot`, I think there's no point having two codepaths, when the fallback works for all. If you're just showcasing texture replacement for fills, then you should use the asset based codepath only.

A different example would be better for showcasing `IsSlot` (perhaps restyling the borders if you have full health or something?)

absoluteAquarian replied:
good point.  i'll think it over

Chicken-Bones commented on ExampleMod/Common/UI/ResourceDisplay/VanillaLifeOverlay.cs:
```diff
@@ -0,0 +1,124 @@
+			bool canReplaceResource;
+
+			if (context.IsSlot(HorizontalBarsPlayerReosurcesDisplaySet.LifePanels)) {
+				// Vanilla "replaces" the bars from right to left, so it would be a good idea to mimic that behaviour
```
this is annoying, can we change the draw order in the vanilla and example mod bars so modders don't need to do this? It fails with the example bars, because we don't check that slot (and in general, may fail with modded resource UIs if they change draw order)

absoluteAquarian replied:
as in, start with the rightmost bar and draw to the left instead of vice versa (the current logic)?

Chicken-Bones commented on ExampleMod/Common/UI/ResourceDisplay/VanillaLifeOverlay.cs:
```diff
@@ -0,0 +1,124 @@
+		public override void PostDrawResource(ResourceOverlayDrawContext context) {
+			// Life resources are drawn over in groups of two
+			// Bars panel drawing has two additional elements, so the "resource index" needs to be offset
+			int fillResourceNumber = context.IsSlot(HorizontalBarsPlayerReosurcesDisplaySet.LifePanels) ? context.resourceNumber - 1 : context.resourceNumber;
```
is there no way to generalise this?

absoluteAquarian replied:
possibly, but the only case where the "max resource number" isn't 20 is only the Bars panels

Chicken-Bones commented on ExampleMod/Common/UI/ResourceDisplay/VanillaLifeOverlay.cs:
```diff
@@ -0,0 +1,124 @@
+
+		private static bool CompareAssetNames(string existingName, string compareAssetPath) {
+			// This is a helper method for checking if a certain vanilla asset was drawn
+			return existingName == Main.Assets.Request<Texture2D>(compareAssetPath).Name;
```
Not recommended to call `Request` on every frame. Cache instead

Chicken-Bones commented on ExampleMod/Common/UI/ResourceDisplay/VanillaLifeOverlay.cs:
```diff
@@ -0,0 +1,124 @@
+
+namespace ExampleMod.Common.UI.ResourceDisplay
+{
+	public class VanillaLifeOverlay : ModResourceOverlay
```
Why does this example need to replace the panel, and not just filling? It seems particularly complicated as a result.

Chicken-Bones commented on ExampleMod/Content/Items/Consumables/ExampleLifeFruit.cs:
```diff
@@ -1,56 +1,52 @@
+				// This handles the 2 achievements related to using any life increasing item or reaching the max amount of consumed Life Crystals, Life Fruit and Mana Crystals.
+				// Ignored since our item is only useable after this achievement is reached
+				// AchievementsHelper.HandleSpecialEvent(player, 2);
+				//TODO re-add this when ModAchievement is merged?
```
remove?

Chicken-Bones commented on ExampleMod/Content/Items/Consumables/ExampleLifeFruit.cs:
```diff
@@ -1,56 +1,52 @@
-			//TODO re-add this when ModAchievement is merged?
-			return true;
+			// Returning null will make the item not be consumed
+			return canConsume ? true : null;
```
just return `true` in the `if`. Better yet, invert the if and early return `null`

Chicken-Bones commented on ExampleMod/Content/Items/Consumables/ExampleManaCrystal.cs:
```diff
@@ -0,0 +1,62 @@
+			}
+
+			// Returning null will make the item not be consumed
+			return canConsume ? true : null;
```
same comments as `ExampleLifeFruit`

Chicken-Bones commented on ExampleMod/ExampleMod.Networking.cs:
```diff
@@ -20,11 +20,12 @@ internal enum MessageType : byte
-					examplePlayer.exampleLifeFruits = reader.ReadInt32();
+					ExampleStatIncreasePlayer examplePlayer = Main.player[playernumber].GetModPlayer<ExampleStatIncreasePlayer>();
+					examplePlayer.exampleLifeFruits = reader.ReadByte();
+					examplePlayer.exampleManaCrystals = reader.ReadByte();
```
Please move into `ExampleStatIncreasePlayer` as an instance method.

Chicken-Bones commented on ExampleMod/ExampleMod.Networking.cs:
```diff
@@ -20,11 +20,12 @@ internal enum MessageType : byte
+					ExampleStatIncreasePlayer examplePlayer = Main.player[playernumber].GetModPlayer<ExampleStatIncreasePlayer>();
+					examplePlayer.exampleLifeFruits = reader.ReadByte();
+					examplePlayer.exampleManaCrystals = reader.ReadByte();
 					// SyncPlayer will be called automatically, so there is no need to forward this data to other clients.
```
It will not as far as I can tell

absoluteAquarian replied:
should a call to SyncPlayer or the like be done here, or should the comment just be removed?

Chicken-Bones replied:
Please test this in multiplayer. I believe you need to forward the packet to all clients except the one it came from if `Main.netMode == NetModeID.Server`

absoluteAquarian replied:
i look again in the code and SyncPlayer is only called automatically in `NetMessage.SyncOnePlayer()` and in MessageBuffer for `MessageID.PlayerInfo` and `MessageID.PlayerSpawn`

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/ResourceSets/ClassicPlayerResourcesDisplaySet.cs.patch:
```diff
@@ -0,0 +1,231 @@
+ 						num5 = 30;
+ 
+-					num6 = num7 / 4f + 0.75f;
++					num6 = (float)(num7 / 4f + 0.75f);
```
this is a lot of difficult patches to maintain :|

absoluteAquarian replied:
`num6` appears to just be used for filling scale due to the "pulse" effect, so it's probably fine to just have `num7` be a `float` variable instead of `double`
same idea for the mana drawing method

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/ResourceSets/PlayerStatsSnapshot.cs.patch:
```diff
@@ -0,0 +1,75 @@
++			numLifeHearts = (int)(LifeMax / num);
++			numManaStars = (int)(ManaMax / 20d);
++
++			SystemLoader.ModifyStatSnapshot(player, ref this);
```
Please remove `ModifyStatSnapshot`, we don't have good use-cases for it yet, and forcing the use of stat snapshot everywhere increases maintenance overhead

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/ResourceSets/ResourceDrawSettings.cs.patch:
```diff
@@ -0,0 +1,61 @@
++				/*
+ 				Vector2 position = topLeftAnchor + drawOffset;
+ 				Vector2 origin = OffsetSpriteAnchor + rectangle.Size() * OffsetSpriteAnchorByTexturePercentile;
++				*/
```
why not move the `DrawResource` hook below these lines, rather than commenting them out.

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/ResourceSets/ResourceDrawSettings.cs.patch:
```diff
@@ -0,0 +1,61 @@
++
++				position = drawContext.position;
++				Vector2 origin = drawContext.origin;
++				Rectangle rectangle = drawContext.source ?? texture.Frame();
```
`drawContext` is not passed by `ref` so there's no way for a mod to alter it, if that was your intent

absoluteAquarian replied:
if formerly was, but i forgot to alter this code after making it not be passed by `ref`

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/ResourceSets/ResourceDrawSettings.cs.patch:
```diff
@@ -0,0 +1,61 @@
+ 		public Vector2 OffsetSpriteAnchorByTexturePercentile;
++		public PlayerStatsSnapshot StatsSnapshot;
++		public IPlayerResourcesDisplaySet DisplaySet;
++		public int ResourceSetSlot;
```
`internal`?

absoluteAquarian replied:
ResourceDrawSettings is public, so if those fields were made internal, modders would be blocked from using the struct properly without resorting to Reflection

Chicken-Bones commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -4703,6 +4724,82 @@
+-			if (sItem.type == 109 && itemAnimation > 0 && statManaMax < 200 && ItemTimeIsZero) {
++			if (sItem.type == 109 && itemAnimation > 0 && ConsumedManaCrystals < ManaCrystalMax && ItemTimeIsZero) {
+ 				ApplyItemTime(sItem);
+-				statManaMax += 20;
+-				statManaMax2 += 20;
+-				statMana += 20;
+-				if (Main.myPlayer == whoAmI)
+-					ManaEffect(20);
```
comment out please

Chicken-Bones commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -5827,6 +5928,31 @@
+ 
+ 							if (player.statMana > 400)
+ 								player.statMana = 400;
++							*/
```
Only one of these you need to comment out is the `statMana`, as the saved values here are vanilla max values anyway, so no problem with capping them to vanilla

Chicken-Bones commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -5690,10 +5787,14 @@
  					using (BinaryWriter binaryWriter = new BinaryWriter(cryptoStream)) {
++						// Keep the player state valid with vanilla (#HealthManaAPI)
++						// Any modifications to statLifeMax/statManaMax by mods can just be done during runtime
++						PlayerLoader.ResetMaxStatsToVanilla(player);
```
This will leave the max stat values incorrect for the rest of the frame, better to just change the saved values in the save function

Chicken-Bones commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -2464,7 +2464,16 @@
+ 			honeyCombItem = null;
+ 			gravControl2 = false;
++			PlayerLoader.ModifyMaxStats(this);
++			/*
+ 			statLifeMax2 = statLifeMax;
+ 			statManaMax2 = statManaMax;
++			*/
```
No need to comment out these two lines.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/PlayerLoader.cs:
```diff
@@ -78,6 +78,44 @@ public static class PlayerLoader
+			player.statLifeMax = (int)cumulativeHealth.ApplyTo(player.statLifeMax);
+			player.statManaMax = (int)cumulativeMana.ApplyTo(player.statManaMax);
+
+			player.statLifeMax2 = player.statLifeMax;
+			player.statManaMax2 = player.statManaMax;
```
Is there a good reason to have this in here, rather than just in `ResetEffects` like vanilla?

absoluteAquarian replied:
they're included in case a modder calls `PlayerLoader.ModifyMaxStats()` for whatever reason

Chicken-Bones replied:
I would rather not have this, better to follow vanilla more closely, and only add this if a modder requests it

Chicken-Bones commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -5827,6 +5928,31 @@
+ 								player.statMana = 400;
++							*/
++
++							PlayerLoader.ModifyMaxStats(player);
```
why this call? `ModPlayer` data hasn't been loaded yet at this point

Chicken-Bones replied:
In fact, this could mess with the vanilla fallback `ConsumedX` calculations if a mod just, for eg, adds a blanket 100hp.

Why not set the `ConsumedX` properties here, instead of using tML NBT data at all, since you'll have the vanilla data in all cases as that's what was saved.

absoluteAquarian replied:
good point

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/SystemLoader.cs:
```diff
@@ -334,6 +335,12 @@ public static partial class SystemLoader
 			}
 		}
 
+		public static void ModifyStatSnapshot(Player player, ref PlayerStatsSnapshot snapshot) {
+			foreach (var system in HookModifyStatSnapshot.arr) {
+				system.ModifyStatSnapshot(player, ref snapshot);
+			}
+		}
```
Remove hook, no good use case yet

absoluteAquarian committed:
removed ModSystem.ModifyStatSnapshot and simplified the overlay examples

absoluteAquarian committed:
implemented two requested changes

- ExampleStatIncrease.SendClientChanges now just calls SyncPlayer
- the example overlays were changed to check asset instances instead of their Name properties

absoluteAquarian committed:
player I/O changes

- removed compound saving in PlayerIO in favor of a better alternative
- added code to save/load the ConsumedX properties directly

absoluteAquarian committed:
removed comments

absoluteAquarian committed:
improved the UseItem examples in ExampleLifeFruit and ExampleManaCrystal

absoluteAquarian committed:
removed unnecessary comment

absoluteAquarian committed:
networking changes

absoluteAquarian committed:
better patch size for ResourceDrawSettings

absoluteAquarian committed:
better patch size for ClassicPlayerResourcesDisplaySet

absoluteAquarian committed:
asset instance caching

absoluteAquarian committed:
better patch size for Player

absoluteAquarian committed:
uncommented statLifeMax and statManaMax capping in Player.LoadPlayer since they're the vanilla values anyway

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModSystem.cs:
```diff
@@ -3,6 +3,7 @@
 using System;
 using System.Collections.Generic;
 using System.IO;
+using Terraria.GameContent.UI.ResourceSets;
```
Unused import now

absoluteAquarian committed:
removed statLifeMax2 and statManaMax2 usage in PlayerLoader.ModifyMaxStats

absoluteAquarian committed:
removed unnecessary using statement

absoluteAquarian committed:
renamed the ResourceDisplay folder to ResourceOverlay

absoluteAquarian committed:
lots of bug fixes and cleanup

- removed `ResourceSlotId` in favor of just checking asset instances
- added a new `ResourceStat` property to `ResourceOverlayDrawContext`
- fixed a bug that caused the display cycling to get stuck between Classic and Swapped Bars
- modified ExampleMod to account for the new changes
- updated XML comments

absoluteAquarian committed:
removed an unnecessary using statement which was causing conflicts

Chicken-Bones commented on ExampleMod/Common/UI/ResourceOverlay/VanillaLifeOverlay.cs:
```diff
@@ -0,0 +1,140 @@
+
+		public override void Unload() {
+			vanillaAssetCache = null;
+		}
```
No need to do this, as it'll get garbage collected fine, the field isn't `static`

Chicken-Bones commented on ExampleMod/Common/UI/ResourceOverlay/VanillaLifeOverlay.cs:
```diff
@@ -0,0 +1,140 @@
+
+		public override void Load() {
+			vanillaAssetCache = new();
+		}
```
toss that in the field declaration

Chicken-Bones commented on ExampleMod/Common/UI/ResourceOverlay/VanillaLifeOverlay.cs:
```diff
@@ -0,0 +1,140 @@
+			bool drawingBarsPanels = CompareAssets(asset, barsFolder + "HP_Panel_Middle");
+
+			// Bars panel drawing has two additional elements, so the "resource index" needs to be offset
+			int fillResourceNumber = drawingBarsPanels ? context.resourceNumber - 1 : context.resourceNumber;
```
This needs to be provided by the `ResourceOverlayDrawContext`, can you showcase using `ResourceStat` instead?

Chicken-Bones commented on ExampleMod/Common/UI/ResourceOverlay/VanillaLifeOverlay.cs:
```diff
@@ -0,0 +1,140 @@
+
+			// "context" contains information used to draw the resource
+			// If you want to draw directly on top of the vanilla hearts, just replace the texture and have the context draw the new texture
+			context.texture = ModContent.Request<Texture2D>(texture, AssetRequestMode.ImmediateLoad);
```
This one should be an `Asset<Texture2D>` field. 
Recommend using `??=`
No need for `ImmediateLoad` (if it doesn't draw for a frame or two, no big loss)

absoluteAquarian replied:
1) fair point.  this is a leftover from the old code that had multiple textures
2) this line replaces the texture variable (which won't be reflected outside of the method) then draws the new texture, hence `=` instead of `??=`

Chicken-Bones replied:
```cs
// field
Asset<Texture2D> panelTexture;

// in method later
context.texture = panelTexture ??= ModContent.Request<Texture2D>(texture);
```

absoluteAquarian replied:
ah, that's what you meant

Chicken-Bones commented on ExampleMod/Common/UI/ResourceOverlay/VanillaLifeOverlay.cs:
```diff
@@ -0,0 +1,140 @@
+
+			if (context.resourceNumber == context.snapshot.AmountOfLifeHearts) {
+				// Final panel to draw has a special "Fancy" variant.  Determine whether it has panels to the left of it
+				if (context.resourceNumber % 10 == 1) {
```
You should check texture name here, rather than `resourceNumber` since that's what the framing info/position actually relates to, and then you wouldn't need to comment `// Vanilla texture is "Heart_Single_Fancy"` for eg

Chicken-Bones commented on ExampleMod/Common/UI/ResourceOverlay/VanillaLifeOverlay.cs:
```diff
@@ -0,0 +1,140 @@
+			context.Draw();
+		}
+
+		private static void DrawFancyPanelOverlay(ResourceOverlayDrawContext context) {
```
Need a comment saying that this is not required, and that the example is of changing the background of the panel when the resource is low.

absoluteAquarian replied:
this is the method responsible for drawing the heart backgrounds

Chicken-Bones replied:
Yes, most people won't need to do custom backgrounds, just fore-grounds.

absoluteAquarian replied:
that's what i meant by "background" -- the part behind the hearts, stars and bar fillings

Chicken-Bones commented on ExampleMod/Common/UI/ResourceOverlay/VanillaLifeOverlay.cs:
```diff
@@ -0,0 +1,140 @@
+			// If you want to draw directly on top of the vanilla bar panels, just replace the texture and have the context draw the new texture
+			context.texture = ModContent.Request<Texture2D>("ExampleMod/Common/UI/ResourceOverlay/BarsLifeOverlay_Panel", AssetRequestMode.ImmediateLoad);
+			// Due to the replacement texture and the vanilla texture having different heights, the source needs to also be modified
+			context.source = (Rectangle)context.source with { Height = context.texture.Height() };
```
Can you use `context.source = context.texture.Frame();` here?

absoluteAquarian replied:
oops, yes, this was mistakenly copied from when i was working on the bar filling drawing

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/ResourceSets/PlayerResourceSetsManager.cs.patch:
```diff
@@ -0,0 +1,48 @@
+ 
+ 		private void Configuration_OnLoad(Preferences obj) {
+-			_activeSetConfigKey = Main.Configuration.Get("PlayerResourcesSet", "New");
++			_activeSetConfigKeyOriginal = _activeSetConfigKey = Main.Configuration.Get("PlayerResourcesSet", "New");
```
Put `_activeSetConfigKeyOriginal = _activeSetConfigKey ` on a new line

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/ResourceSets/PlayerResourceSetsManager.cs.patch:
```diff
@@ -0,0 +1,48 @@
+ 
+ 		public void LoadContent(AssetRequestMode mode) {
+ 			_sets["New"] = new FancyClassicPlayerResourcesDisplaySet("New", "New", "FancyClassic", mode);
+-			_sets["Default"] = new ClassicPlayerResourcesDisplaySet("Default", "Default");
+ 			_sets["HorizontalBars"] = new HorizontalBarsPlayerReosurcesDisplaySet("HorizontalBars", "HorizontalBars", "HorizontalBars", mode);
++			_sets["Default"] = new ClassicPlayerResourcesDisplaySet("Default", "Default");
```
why re-order?

Chicken-Bones replied:
The values of a dictionary aren't ordered anyway.

absoluteAquarian replied:
they're iterated through in the order they're added at when using FirstOrDefault

Chicken-Bones replied:
Well I looked at the reference source, and apparently it's true for this version of .NET

I don't like relying on it though, it's confusing to other developers who follow the typical advice (that dictionaries and hashmaps are unordered)

From MSDN
![image](https://user-images.githubusercontent.com/388233/191880518-7efba9ce-cde0-4adb-8a88-725b97636dcf.png)

absoluteAquarian replied:
fair point, though that's what the vanilla logic uses.  i should mention that `Dictionary<K, V>.Values` is what's being iterated

Chicken-Bones replied:
Yeah, the version of `.NET` we're using stores all the entries in an `Entry[]` in insertion order (at least when you don't remove and re-add stuff)

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/ResourceSets/PlayerStatsSnapshot.cs.patch:
```diff
@@ -0,0 +1,73 @@
+ 		public int ManaMax;
+-		public float ManaPerSegment;
++		// Made double to prevent bugs caused by precision limits (#HealthManaAPI)
++		public double ManaPerSegment => ManaMax / (double)AmountOfManaStars;
```
what was wrong with `float` here? Can you give an example?

absoluteAquarian replied:
Let's take the example I used in the Discord thread:
1) `LifeMax` is set to `290` and `AmountOfLifeHearts` is `15` (e.g. a mod adds a "negative heart" which removes 10 max life)
2) `290 / 15d = 19.333333333333332d` and `290 / 15f = 19.333334f`
3) code that then tries to get the amount of hearts from `LifeMax / AmountOfLifeHearts` or `LifePerHeart` for whatever reason ends up with `15.000000000000002d` -- which gets correctly truncated to `15` -- instead of `14.999999f` -- which gets incorrectly truncated to `14`

Chicken-Bones replied:
This issue is not exclusive to `float` over double. This can happen any time you try to do `v = a / b` and then expect  `a / v == b`

```cs
for (int i = 200; i < 400; i++) { 
     double d = i / 15d;
     int k = (int)(i / d);
     if (k != 15) Console.WriteLine(i);
}
```
```
> 295/15d
19.666666666666668
> 295/(295/15d)
14.999999999999998
```

absoluteAquarian replied:
right.  i'll revert the `float`->`double` changes

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/ResourceSets/HorizontalBarsPlayerReosurcesDisplaySet.cs.patch:
```diff
@@ -0,0 +1,177 @@
+ 
+ 		private void LifeFillingDrawer(int elementIndex, int firstElementIndex, int lastElementIndex, out Asset<Texture2D> sprite, out Vector2 offset, out float drawScale, out Rectangle? sourceRect) {
+ 			sprite = _hpFill;
+-			if (elementIndex >= _hpSegmentsCount - _hpFruitCount)
++			// Make the filling draw from right to left (#HealthManaAPI)
++			if (elementIndex < _hpFruitCount)
+ 				sprite = _hpFillHoney;
+ 
+ 			FillBarByValues(elementIndex, sprite, _hpSegmentsCount, _hpPercent, out offset, out drawScale, out sourceRect);
++			// Make the bar fillings draw from right to left (#HealthManaAPI)
++			int opposite = lastElementIndex - (elementIndex - firstElementIndex);
++			int drawIndexOffset = opposite - elementIndex;
++			offset.X += drawIndexOffset * sprite.Width();
+ 		}
+ 
+-		private static void FillBarByValues(int elementIndex, Asset<Texture2D> sprite, int segmentsCount, float fillPercent, out Vector2 offset, out float drawScale, out Rectangle? sourceRect) {
++		public static void FillBarByValues(int elementIndex, Asset<Texture2D> sprite, int segmentsCount, float fillPercent, out Vector2 offset, out float drawScale, out Rectangle? sourceRect) {
+ 			sourceRect = null;
+ 			offset = Vector2.Zero;
+ 			float num = 1f;
+ 			float num2 = 1f / (float)segmentsCount;
+-			float t = 1f - fillPercent;
++			float t = fillPercent;
+ 			float lerpValue = Utils.GetLerpValue(num2 * (float)elementIndex, num2 * (float)(elementIndex + 1), t, clamped: true);
+-			num = 1f - lerpValue;
++			num = lerpValue;
```
Please comment out the vanilla code here rather than removing it

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/ResourceSets/ClassicPlayerResourcesDisplaySet.cs.patch:
```diff
@@ -0,0 +1,226 @@
+-			if (mouseScreen.X > (float)(762 + UI_ScreenAnchorX) && mouseScreen.X < (float)(762 + num + UI_ScreenAnchorX) && mouseScreen.Y > 30f && mouseScreen.Y < (float)(30 + num2))
++			num2 = 28 * snapshot.AmountOfManaStars;
++			if (mouseScreen.X > (float)(762 + UI_ScreenAnchorX) && mouseScreen.X < (float)(762 + num + UI_ScreenAnchorX) && mouseScreen.Y > 30f && mouseScreen.Y < (float)(30 + num2) && ResourceOverlayLoader.DisplayHoverText(snapshot, this, false))
+ 				CommonResourceBarMethods.DrawManaMouseOver();
```
Is there any way we can reduce the patch size of this file? There's so much here that could break in an update

Chicken-Bones replied:
I guess you don't need to use `snapshot` everywhere now that the hook is removed?

absoluteAquarian replied:
yeah, i could safely revert the `snapshot` changes in the Classic display set due to the hook not being present anymore

absoluteAquarian committed:
ExampleMod changes + reduced patch file sizes

absoluteAquarian committed:
reduced patch size

absoluteAquarian committed:
added comments indicating that drawing over the panels isn't required

absoluteAquarian committed:
more asset caching

absoluteAquarian committed:
removed `ResourceStat` from `ResourceOverlayDrawContext` and its related fields in `ResourceDrawSettings`

absoluteAquarian committed:
reduced patch sizes

absoluteAquarian committed:
added improved setting and cycling logic for the display sets

Chicken-Bones commented on ExampleMod/Common/UI/ResourceOverlay/VanillaLifeOverlay.cs:
```diff
@@ -0,0 +1,138 @@
+			bool drawingBarsPanels = CompareAssets(asset, barsFolder + "HP_Panel_Middle");
+
+			// Bars panel drawing has two additional elements, so the "resource index" needs to be offset
+			int fillResourceNumber = drawingBarsPanels ? context.resourceNumber - 1 : context.resourceNumber;
```
Alright, all we need is for this value to be provided by the context, rather than calculated here.

Chicken-Bones reviewed (changes requested):
Incredibly, I think this is the only change still required

absoluteAquarian committed:
API changes and bug fixes

- made `ResourceOverlayDrawContext.resourceNumber` start at 0 by default
- made the Bars panel drawing have `resourceNumber` start at -1
- added `ResourceIndexOffset` to `ResourceDrawSettings`
- actually tested the changed display set cycling code
- various bug fixes
